File: src/test/java/com/bigdata/sort/M9BucketSort.java
package com.bigdata.sort;

import org.junit.Test;

import java.util.ArrayList;

import java.util.List;

/**
 * 桶排序
 * 1.找到数组最大、最下值，得到数据分部范围；
 * 2.基于数据量选择合适分桶数，组件分桶数组；
 * 3.从最小桶开始收集数据，收集满后进行排序，然输出到中间数组
 * 4.中间数组拷贝到原数组
 * 时间复杂度 o(n+k), 空间复杂度 o(n+k) 稳定
 */
public class M9BucketSort extends BaseSort{

    @Test
    public void testSolve(){
        solve1(arr);
    }

    public static void solve1(int[] arr){
        // 根据数据质量，决定分桶数
        int bucketNum = getbucketNum(arr.length);
        // 根据分桶数划分桶边界数组
        int[] buckets = getBuckets(arr,bucketNum);
        // 顺序接收排好序的桶数据
        int[] result = new int[arr.length];
        //
        int j=0;
        for(int i=1;i<buckets.length;i++){
            int left = buckets[i-1];
            int right = buckets[i];
            int[] sortedBucket = enterBucket(arr, left, right);
            System.arraycopy(sortedBucket,0,result,j,sortedBucket.length);
            j += sortedBucket.length;
        }
        System.arraycopy(result,0,arr,0,arr.length);
    }

    // 按起止边界抽取数据入桶，并对桶内数据进行排序，然后输出
    public static int[] enterBucket(int[] arr,int left,int right){
        List<Integer> temp = new ArrayList<>();
        for(int num:arr){
            if(num>=left && num < right){
                temp.add(num);
            }
        }
        int[] bucket = new int[temp.size()];
        for(int i=0;i<bucket.length;i++){
            bucket[i] = temp.get(i);
        }
        insertSort(bucket);
        return bucket;
    }

    // 根据数据量决定分桶数
    public static int getbucketNum(int length){
        int num = 0;
        if(length<100){
            num = 2;
        }else if(length<1000){
             num = 5;
        }else if(length<10000){
            num = 8;
        }else{
            num = 10;
        }
        return num;
    }

    // 获取分桶边界数组
    public static int[] getBuckets(int[] arr,int bucketNum){
        int minPos = 0;
        int maxPos = 0;
        for(int i=0;i<arr.length;i++){
            if(arr[i]<arr[minPos]){
                minPos = i;
            }
            if(arr[i]>arr[maxPos]){
                maxPos = i;
            }
        }

        int len = arr.length % bucketNum !=0 ? (arr.length / bucketNum) +1 : arr.length / bucketNum;
        int[] result = new int[len];
        result[0] = arr[minPos];
        result[len-1] = arr[maxPos]+1;
        for(int i=1;i<result.length-1;i++){
            result[i] = result[i-1] + bucketNum;
        }
        return result;
    }

    // 桶内部使用插入排序
    public static void insertSort(int[] arr){
//        Arrays.sort(arr);
        for(int i=1;i<arr.length;i++){
            for(int j=i-1;j>=0;j--){
                if(arr[j]>arr[j+1]){
                    swap(arr,j,j+1);
                }
            }
        }
    }

}

File: python2/sort/M9BucketSort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os, sys


# 桶排序
# 1.找到数组最大、最下值，得到数据分部范围；
# 2.基于数据量选择合适分桶数，组件分桶数组；
# 3.从最小桶开始收集数据，收集满后进行排序，然输出到中间数组
# 4.中间数组拷贝到原数组
# 时间复杂度 o(n+k), 空间复杂度 o(n+k) 稳定

class M9BucketSort:

    @staticmethod
    def solve1(arr):
        num = M9BucketSort.get_boucket_num(len(arr))
        min,max = M9BucketSort.get_boundary(arr)
        buckets = M9BucketSort.get_buckets(num, min, max)
        temp = []
        for i in range(len(buckets)-1):
            sorted_bucket = M9BucketSort.entry_bucket(arr, buckets[i], buckets[i + 1])
            temp.extend(sorted_bucket)
        M9BucketSort.copy_array(temp,arr)

    @staticmethod
    def copy_array(src,dest):
        for i in range(len(dest)):
            dest[i] = src[i]

    @staticmethod
    def entry_bucket(arr,left,right):
        temp = []
        for e in arr:
            if e>=left and e < right:
                temp.append(e)
        M9BucketSort.sort(temp)
        return temp

    @staticmethod
    def sort(arr):
        for i in range(1,len(arr)):
            for j in range(i-1,-1,-1):
                if arr[j] > arr[j+1]:
                    M9BucketSort.swap(arr,j,j+1)

    @staticmethod
    def swap(arr,i,j):
        arr[i],arr[j] = arr[j],arr[i]

    @staticmethod
    def get_buckets(num,min,max):
        width = max - min
        slide = width / num
        buckets = [0] * (num+1)
        buckets[0] = min
        buckets[-1] = max + 1
        for i in range(1,len(buckets)-1):
            buckets[i] = buckets[i-1] + slide
        return buckets

    @staticmethod
    def get_boucket_num(length):
        num = 0
        if length<100:
            num  = 2
        elif length<1000:
            num = 4
        else:
            num = 6
        return num

    @staticmethod
    def get_boundary(arr):
        min_pos = 0
        max_pos = 0
        for i in range(len(arr)):
            if arr[i]<arr[min_pos]:
                min_pos = i
            if arr[i]>arr[max_pos]:
                max_pos = i
        return [arr[min_pos],arr[max_pos]]

def test_solve():
    arr = [-10, -9, 3, 2, 5, 10, 20]
    print(arr)
    M9BucketSort.solve1(arr)
    print(arr)

if __name__ == "__main__":
    test_solve()
    pass